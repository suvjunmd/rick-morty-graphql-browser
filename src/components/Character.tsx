
import { useQuery, gql, ApolloError } from "@apollo/client";

interface CharacterProps {
  id: string;
}

interface QueryResponse {
  loading: boolean;
  error?: ApolloError | undefined;
  data?: QueryData;
}

interface QueryData {
  character: QueryCharacter;
}

interface QueryCharacter {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: CharacterOrigin;
  location: CharacterLocation;
  image: string;
  created: string;
}

interface CharacterOrigin {
  name: string;
}

interface CharacterLocation {
  name: string;
}

export const GET_CHARACTER_QUERY = gql`
  query GetCharacter($id: ID!) {
    character(id: $id) {
      id
      name
      status
      species
      type
      gender
      origin {
        name
      }
      location {
        name
      }
      image
      created
    }
  }
`;

export default function Character({ id }: CharacterProps) {
  const queryResponse: QueryResponse = useQuery(GET_CHARACTER_QUERY, {
    variables: { id },
  });

  const { loading, error, data } = queryResponse;

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error : {error.message}</p>;
  if (!data) return <p>No results</p>;

  return (
    <div>
      <h3>
        {data.character.id} - {data.character.name}
      </h3>
      <img width="300" height="300" alt="" src={`${data.character.image}`} />
      <p>
        <b>ID:</b> {data.character.id}
      </p>
      <p>
        <b>Name:</b> {data.character.name}
      </p>
      <p>
        <b>Status:</b> {data.character.status}
      </p>
      <p>
        <b>Species:</b> {data.character.species}
      </p>
      <p>
        <b>Type:</b> {data.character.type}
      </p>
      <p>
        <b>Gender:</b> {data.character.gender}
      </p>
      <p>
        <b>Origin:</b> {data.character.origin.name}
      </p>
      <p>
        <b>Location:</b> {data.character.location.name}
      </p>
      <p>
        <b>Created:</b> {new Date(data.character.created).toLocaleString()}
      </p>
    </div>
  );
}
