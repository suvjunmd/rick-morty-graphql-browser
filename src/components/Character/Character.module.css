.container {
  max-width: 100%;
  width: fit-content;
  margin-left: auto;
  margin-right: auto;
}

.name {
  text-align: center;
}
