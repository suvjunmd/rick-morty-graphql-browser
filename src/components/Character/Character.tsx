import { useQuery } from "@apollo/client";
import styles from "./Character.module.css";
import { graphql } from "../../gql";

interface CharacterProps {
  id: string | undefined;
}

export const GET_CHARACTER_QUERY = graphql(`
  query GetCharacter($id: ID!) {
    character(id: $id) {
      id
      name
      status
      species
      type
      gender
      origin {
        name
      }
      location {
        name
      }
      image
      created
    }
  }
`);

export default function Character({ id }: CharacterProps) {
  if (!id) return <p>No results</p>;

  const { loading, error, data } = useQuery(GET_CHARACTER_QUERY, {
    variables: { id },
  });

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error : {error.message}</p>;
  if (!data?.character) return <p>No results</p>;

  return (
    <div className={styles.container}>
      <h1 className={styles.name}>{data.character?.name}</h1>
      <img width="300" height="300" alt="" src={`${data.character.image}`} />
      <p>
        <b>ID:</b> {data.character.id}
      </p>
      <p>
        <b>Status:</b> {data.character.status}
      </p>
      <p>
        <b>Species:</b> {data.character.species}
      </p>
      <p>
        <b>Type:</b> {data.character.type}
      </p>
      <p>
        <b>Gender:</b> {data.character.gender}
      </p>
      <p>
        <b>Origin:</b> {data.character.origin?.name}
      </p>
      <p>
        <b>Location:</b> {data.character.location?.name}
      </p>
      <p>
        <b>Created:</b>{" "}
        {data.character.created &&
          new Date(data.character.created).toLocaleString()}
      </p>
    </div>
  );
}
