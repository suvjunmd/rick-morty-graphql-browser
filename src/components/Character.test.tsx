import { expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import Character, { GET_CHARACTER_QUERY } from "./Character";

const mocks = [
  {
    request: {
      query: GET_CHARACTER_QUERY,
      variables: {
        id: "1",
      },
    },
    result: {
      data: {
        character: {
          id: "1",
          name: "Rick",
          status: "Alive",
          species: "Human",
          type: "",
          gender: "Male",
          origin: {
            name: "Earth (C-137)",
          },
          location: {
            name: "Citadel of Ricks",
          },
          image: "1.jpeg",
          created: "2022-11-04T18:48:46.250Z",
        },
      },
    },
  },
];

it("should render character", async () => {
  const { container } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Character id="1" />
    </MockedProvider>
  );
  expect(await screen.findByText("Loading...")).toBeInTheDocument();
  expect(await screen.findByText("Citadel of Ricks")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

it("should render no results message", async () => {
  const noResultsMock = {
    request: {
      query: GET_CHARACTER_QUERY,
      variables: {
        id: "1",
      },
    },
    result: {},
  };
  render(
    <MockedProvider mocks={[noResultsMock]} addTypename={false}>
      <Character id="1" />
    </MockedProvider>
  );
  expect(await screen.findByText("No results")).toBeInTheDocument();
});

it("should render network error message", async () => {
  const errorMock = {
    request: {
      query: GET_CHARACTER_QUERY,
      variables: {
        id: "1",
      },
    },
    error: new Error("An error occurred"),
  };
  render(
    <MockedProvider mocks={[errorMock]} addTypename={false}>
      <Character id="1" />
    </MockedProvider>
  );
  expect(
    await screen.findByText("Error : An error occurred")
  ).toBeInTheDocument();
});
