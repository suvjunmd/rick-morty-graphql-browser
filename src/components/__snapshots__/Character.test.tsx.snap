// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render character 1`] = `
<div>
  <div
    class="character-container"
  >
    <h1>
      Rick
    </h1>
    <img
      alt=""
      height="300"
      src="1.jpeg"
      width="300"
    />
    <p>
      <b>
        ID:
      </b>
       
      1
    </p>
    <p>
      <b>
        Status:
      </b>
       
      Alive
    </p>
    <p>
      <b>
        Species:
      </b>
       
      Human
    </p>
    <p>
      <b>
        Type:
      </b>
       
    </p>
    <p>
      <b>
        Gender:
      </b>
       
      Male
    </p>
    <p>
      <b>
        Origin:
      </b>
       
      Earth (C-137)
    </p>
    <p>
      <b>
        Location:
      </b>
       
      Citadel of Ricks
    </p>
    <p>
      <b>
        Created:
      </b>
       
      11/4/2022, 8:48:46 PM
    </p>
  </div>
</div>
`;
